### PR TITLE
[SPARK-34892][SS] Introduce MergingSortWithSessionWindowStateIterator sorting input rows and rows in state efficiently

### PR DESCRIPTION
**Pull request by HeartSaVioR**
Introduction: this PR is a part of SPARK-10816 (EventTime based sessionization (session window)). Please refer #31937 to see the overall view of the code change. (Note that code diff could be diverged a bit.)

### What changes were proposed in this pull request?

This PR introduces MergingSortWithSessionWindowStateIterator, which does "merge sort" between input rows and sessions in state based on group key and session's start time. 

Note that the iterator does merge sort among input rows and sessions grouped by grouping key. The iterator doesn't provide sessions in state which keys don't exist in input rows. For input rows, the iterator will provide all rows regardless of the existence of matching sessions in state.

MergingSortWithSessionWindowStateIterator works on the precondition that given iterator is sorted by "group keys + start time of session window", and the iterator still retains the characteristic of the sort.

### Why are the changes needed?

This part is a one of required on implementing SPARK-10816.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT added.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Same here; I marked this as draft as other PRs has to be reviewed and merged earlier. I'll rebase this PR once all other PRs are merged.

**Comment by HeartSaVioR** [comment]
cc. @viirya @xuanyuanking Please take a look. Thanks!

**Comment by viirya** [comment]
Thanks @HeartSaVioR. I will review this today.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala:71]: A few suggestions:

currentRow -> currentRowFromInput
currentStateRow -> currentRowFromState
currentStateIter -> sessionIterFromState
currentStateFetchedKey -> currentSessionKey

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala:100]: Does this case mean, the input iterator advances to new keys other than current sessions from the state? So we should output from current sessions until it ends and retrieves new sessions from the state again?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIteratorSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIteratorSuite.scala:104]: Hm? Why it is empty? If input is empty, doesn't it output sorted sessions in the state?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIteratorSuite.scala:104]: Oh, nvm. I see.

**Review by viirya — COMMENTED** [review]
Looks good. I have a few suggestions about variable names.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala:71]: And maybe add a few comments explaining these variables for readability.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MergingSortWithSessionWindowStateIterator.scala:100]: Exactly. We retrieve state rows for the key only when there's a new key from input side, so the case is not possible state side advances compared to input side. If the keys differ, either state side also doesn't have any rows to process, or there're rows to process in state side.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
I will merge this tomorrow. Thanks @HeartSaVioR

**Comment by HeartSaVioR** [comment]
Please bear with me on merging this one now to unblock the last PR #33081. We can do post-review for any part of session window changes even it's merged during QA period (and even RC period if someone catches defect.)

**Comment by HeartSaVioR** [comment]
Thanks for reviewing! I merged this to master/3.2. I'll rebase #33081 to reflect the latest master branch.

**Pull request closed without merging**
